L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,390 @@
+L46: C-core: New TLS Credentials API
+----
+* Author(s): gtcooke94, ZhenLian
+* Approver: 
+* Status: Draft
+* Implemented in:
+* Last updated: May 16, 2024
+
+## Abstract
+
+This proposal aims to provide a new TLS credentials API with the following features:
+1. Credential reloading: reload transport credentials periodically from various sources, e.g. from credential files, or users' customized Provider implementations, without shutting down.
+1. Customizable verification of the peer certificate: perform custom validation checks on the peer's certificate chain after it has been verified to chain up to a root of trust, or fully customize path building and cryptographic verification.
+1. TLS version configuration: optionally configure the minimum and maximum of the TLS versions to be negotiated during the handshake.
+1. Support for certificate revocation via Certificate Revocation Lists (CRLs).
+
+## Background
+
+The current TLS implementation (the SSL credentials API) in gRPC C core has some restrictions:
+1. Lack of the credential reloading support. Once an application starts, it can never reload its identity key/cert-chain and root certificates without shutting down.
+1. Inflexibility in the peer verification. There is no API to add additional custom verification checks.
+1. Inability to choose the maximum/minimum TLS version to be negotiated during the TLS handshake.
+1. No support for certificate revocation.
+
+We propose an API that will meet the following requirements:  
+
+1. Flexible enough to support multiple use cases. Users should be able to write their own reloading or verification logic if the provided ones do not fit their needs.
+1. The credential reloading should be efficient. We shall reload only when needed rather than reloading per TLS connection.
+1. C core API needs to be simple enough, so that most of existing SSL credentials in wrapped languages can migrate to call this new API.
+
+### Related Proposals:
+
+1. [L46: Create a new TLS credential API that supports SPIFFE mutual TLS](https://github.com/grpc/proposal/pull/98) - this was subsumed into this proposal.
+1. [A69: Crl Providers](https://github.com/grpc/proposal/pull/382)
+
+## Proposal
+
+### TLS credentials and TLS Options
+This part of the proposal introduces `TlsCredentialsOptions` and TLS credentials. Users use `TlsChannelCredentialsOptions` and `TlsServerCredentialsOptions` to configure the specific security properties they want, and build the client/server credential by passing in those options.
+
+```c++
+// Base class of configurable options specified by users to configure their
+// certain security features supported in TLS. It is used for experimental
+// purposes for now and it is subject to change.
+class TlsCredentialsOptions {
+ public:
+  // Constructor for base class TlsCredentialsOptions.
+  //
+  // @param certificate_provider the provider which fetches TLS credentials that
```

Comment:
Should we make the constructor protected instead of public, so that this class is not instantiable by users? Note that removing the constructor just means that only the default constructor is available, IIRC.

Also, the real class seems to have a public constructor - are we planning on